adaptive/tests/test_learners.py

```diff
@@ -694,3 +695,60 @@ def test_learner_subdomain(learner_type, f, learner_kwargs):
     perform 'similarly' to learners defined on that subdomain only."""
     # XXX: not sure how to implement this. How do we measure "performance"?
     raise NotImplementedError()
+
+
+@run_with(
+    Learner1D,
+    Learner2D,
+    LearnerND,
+    AverageLearner,
+    AverageLearner1D,
+    SequenceLearner,
+)
+def test_to_dataframe(learner_type, f, learner_kwargs):
+    if learner_type is LearnerND:
+        kw = {"point_names": tuple("xyz")[: len(learner_kwargs["bounds"])]}
+    else:
+        kw = {}
+
+    learner = learner_type(generate_random_parametrization(f), **learner_kwargs)
+
+    # Test empty dataframe
+    df = learner.to_dataframe(**kw)
+    assert len(df) == 0
+
+    # Run the learner
+    simple_run(learner, 100)
+    df = learner.to_dataframe(**kw)
+    assert isinstance(df, pandas.DataFrame)
+    if learner_type is AverageLearner1D:
+        assert len(df) == learner.nsamples
+    else:
+        assert len(df) == learner.npoints
+
+    # Add points from the DataFrame to a new empty learner
+    learner2 = learner_type(learner.function, **learner_kwargs)
+    learner2.load_dataframe(df, **kw)
+    assert learner2.npoints == learner.npoints
+
+    # Test this for a learner in a BalancingLearner
+    learners = [
+        learner_type(generate_random_parametrization(f), **learner_kwargs)
+        for _ in range(2)
+    ]
+    bal_learner = BalancingLearner(learners)
+    simple_run(bal_learner, 100)
+    df_bal = bal_learner.to_dataframe(**kw)
+    assert isinstance(df_bal, pandas.DataFrame)
```

Comment:
What happens if you mix learner types in a balancing learner?

Reply:
We do not support that:
https://github.com/python-adaptive/adaptive/blob/c59405ac12f2422c0e6b014ab76fd7683984556f/adaptive/learner/balancing_learner.py#L84-L87